File: ta_lib/strategies/signal/src/zerocross/roc.rs
use base::prelude::*;
use core::prelude::*;
use momentum::roc;
use timeseries::prelude::*;

pub struct RocZeroCrossSignal {
    source: SourceType,
    period: usize,
}

impl RocZeroCrossSignal {
    pub fn new(source: SourceType, period: f32) -> Self {
        Self {
            source,
            period: period as usize,
        }
    }
}

impl Signal for RocZeroCrossSignal {
    fn lookback(&self) -> usize {
        self.period
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let roc = roc(&data.source(self.source), self.period);

        (roc.cross_over(&ZERO), roc.cross_under(&ZERO))
    }
}

File: factor/signal/zerocross/roc.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from core.models.source import SourceType
from factor.signal.base import Signal, SignalType


@dataclass(frozen=True)
class RocZeroCrossSignal(Signal):
    type: SignalType = SignalType.RocZeroCross
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    period: Parameter = StaticParameter(9.0)
